lib/cloud_controller/upload_buildpack.rb

```diff
@@ -47,8 +50,31 @@ def upload_buildpack(buildpack, bits_file_path, new_filename)
       true
     end
 
+    def extract_stack_from_buildpack(bits_file_path)
+      bits_file_path = bits_file_path.path if bits_file_path.respond_to?(:path)
+      output, _, status = Open3.capture3('unzip', '-p', bits_file_path, 'manifest.yml')
+      YAML.safe_load(output).dig('stack') if status.success?
+    end
+
     private
 
+    def determine_new_stack(buildpack, bits_file_path)
+      extracted_stack = extract_stack_from_buildpack(bits_file_path)
+      if extracted_stack.to_s != '' && !Stack.where(name: extracted_stack).first
```

Comment:
`extracted_stack.to_s != ''` can be `extracted_stack.present?`

Reply:
`!Stack.where(name: extracted_stack).first` can be `!Stack.where(name: extracted_stack).empty?` (this will stop Sequel from loading the whole model)

Reply:
Think you meant `Stack.where(name: extracted_stack).empty?` without the `!` but otherwise definitely makes sense.